File: python/src/httpstate/httpstate.py
# HTTPState, https://httpstate.com/
# Copyright (C) Alex Morales, 2026

# Unless otherwise stated in particular files or directories, this software is free software.
# You can redistribute it and/or modify it under the terms of the GNU Affero
# General Public License as published by the Free Software Foundation, either
# version 3 of the License, or (at your option) any later version.

import asyncio
import datetime
import json
import struct
import threading
import urllib.error
import urllib.request
import websockets

from typing import Callable, Dict, List

def get(uuid:str, args:None|Dict = None) -> None|str|dict:
  try:
    req:urllib.request.Request = urllib.request.Request(f'https://httpstate.com/{uuid}')

    if args and args.get('Authorization'):
      req.add_header('Authorization', args.get('Authorization'))

    with urllib.request.urlopen(req) as f:
      if f.status == 200:
        data:str = f.read().decode('utf-8', 'replace')

        if(
             not args
          or (
                not args.get('ETag')
            and not args.get('Last-Modified')
          )
        ):
          return data
        else:
          return {
            **({ 'ETag':f.headers.get('ETag') } if args.get('ETag') else {}),
            **({ 'Last-Modified':f.headers.get('Last-Modified') } if args.get('Last-Modified') else {}),
            'data':data
          }
  except urllib.error.HTTPError as e:
    if e.code == 401:
      raise Exception('401 Unauthorized')
    elif e.code == 404:
      raise Exception('404 Not Found')
    elif e.code == 429:
      raise Exception('429 Too Many Requests')
  except Exception as e:
    print(datetime.datetime.now().isoformat(), 'get.error', e)
    
    raise e

class MessageType:
  def __init__(self, uuid:str, timestamp:int, type:int, value:bytes) -> None:
    self.uuid:str = uuid
    self.timestamp:int = timestamp
    self.type:int = type
    self.value:bytes = value

class Message:
  @staticmethod
  def unpack(b:bytes) -> None|MessageType:
    header:int = b[0]

    if header == 0:
      length:int = b[1]

      return MessageType(
        uuid=b[2:2+length].decode('utf-8'),
        timestamp=struct.unpack_from('>Q', b, 2+length)[0],
        type=b[2+length+8],
        value=b[2+length+9:]
      )

message:type = Message

def post(uuid:str, data:None|str = None, args:None|Dict = None) -> None|int:
  return set(uuid, data, args)

def put(uuid:str, data:None|str = None, args:None|Dict = None) -> None|int:
  return set(uuid, data, args)

def read(uuid:str, args:None|Dict = None) -> None|str|dict:
  return get(uuid, args)

def set(uuid:str, data:None|str = None, args:None|Dict = None) -> None|int:
  if(data is None):
    data = ''

  headers:Dict[str, str] = { 'Content-Type':'text/plain;charset=UTF-8' }

  if args and args.get('Authorization'):
    headers['Authorization'] = args.get('Authorization')

  req:urllib.request.Request = urllib.request.Request(
    f'https://httpstate.com/{uuid}',
    data=data.encode('utf-8'),
    headers=headers,
    method='POST'
  )

  try:
    with urllib.request.urlopen(req) as f:
      return f.status
  except urllib.error.HTTPError as e:
    if e.code == 401:
      raise Exception('401 Unauthorized')
    elif e.code == 404:
      raise Exception('404 Not Found')
    elif e.code == 413:
      raise Exception('413 Content Too Large')

    return e.code
  except Exception as e:
    print(datetime.datetime.now().isoformat(), 'set.error', e)

def write(uuid:str, data:None|str = None, args:None|Dict = None) -> None|int:
  return set(uuid, data, args)

# HTTPState
class HttpState:
  def __init__(self, uuid:str, args:None|Dict = None) -> None:
    self.authorization:None|str = args.get('Authorization') if args else None
    self.data:None|str = None
    self.el:None|asyncio.AbstractEventLoop = None
    self.et:None|Dict[str, List[Callable[[None|str], None]]] = {}
    self.lock:None|threading.Lock = threading.Lock()
    self.uuid:None|str = uuid
    self.ws:None|websockets.WebSocketClientProtocol = None

    threading.Thread(
      daemon=True,
      target=self._el
    ).start()

    threading.Thread(
      daemon=True,
      target=lambda : asyncio.run(self._ws())
    ).start()
  
  def _el(self) -> None:
    self.el:asyncio.AbstractEventLoop = asyncio.new_event_loop()
    
    asyncio.set_event_loop(self.el)
    
    self.el.call_soon_threadsafe(lambda : self.get())

    self.el.run_forever()

  async def _ws(self) -> None:
    self.ws:websockets.WebSocketClientProtocol = await websockets.connect(f"wss://httpstate.com/{self.uuid}")

    await self.ws.send(json.dumps({ 'open':self.uuid, **({ 'Authorization':self.authorization } if self.authorization is not None else {}) }))

    self.emit('open')

    async def data() -> None:
      async for data in self.ws:
        data:None|MessageType = message.unpack(data)

        if(
              data
          and data.uuid == self.uuid
          and data.type == 1
        ):
          with self.lock:
            self.data:None|str = data.value.decode()

          self.emit('change', self.data)
    
    asyncio.create_task(data())

    async def interval() -> None:
      while True:
        try:
          await self.ws.ping()

          await asyncio.sleep(30) # 30 SECONDS
        except websockets.ConnectionClosed:
          break

    asyncio.create_task(interval())
    
    await asyncio.Event().wait()
  
  def delete(self) -> None:
    if self.el is not None:
      if self.ws is not None:
        asyncio.run_coroutine_threadsafe(self.ws.close(), self.el)

      self.el.call_soon_threadsafe(self.el.stop)

    self.authorization = None
    self.data = None
    self.el = None
    self.et = None
    self.lock = None
    self.uuid = None
    self.ws = None

  def emit(self, type:str, data:None|str = None) -> "HttpState":
    if self.et is not None:
      for callback in self.et.get(type, []):
        if(data is None):
          callback()
        else:
          callback(data)

    return self

  def get(self) -> None|str:
    args:None|Dict = { 'Authorization':self.authorization } if self.authorization is not None else None
    data:None|str = get(self.uuid, args)

    with self.lock:
      if(data != self.data):
        if self.el is not None:
          self.el.call_soon_threadsafe(lambda : self.emit('change', self.data))

      self.data = data

    return self.data

  def off(self, type:str, callback:Callable[[None|str], None]) -> "HttpState":
    if self.et is not None and type in self.et:
      try:
        self.et[type].remove(callback)
      except ValueError:
        pass

      if not self.et[type]:
        del self.et[type]

    return self

  def on(self, type:str, callback:Callable[[None|str], None]) -> "HttpState":
    if self.et is not None:
      if type not in self.et:
        self.et[type] = []

      self.et[type].append(callback)

    return self
  
  def post(self, data:None|str = None) -> None|int:
    return self.set(data)
  
  def put(self, data:None|str = None) -> None|int:
    return self.set(data)

  def read(self) -> None|str:
    return self.get()

  def set(self, data:None|str = None) -> None|int:
    args:None|Dict = { 'Authorization':self.authorization } if self.authorization is not None else None
    return set(self.uuid, data, args)

  def write(self, data:None|str = None) -> None|int:
    return self.set(data)

File: java/src/com/httpstate/HttpState.java
// HTTP State, https://httpstate.com/
// Copyright (C) Alex Morales, 2026
//
// Unless otherwise stated in particular files or directories, this software is free software.
// You can redistribute it and/or modify it under the terms of the GNU Affero
// General Public License as published by the Free Software Foundation, either
// version 3 of the License, or (at your option) any later version.

package com.httpstate;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.WebSocket;
import java.net.http.WebSocket.Listener;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

public class HttpState {
  static String Get(String uuid) throws Exception {
    HttpClient httpClient = HttpClient.newHttpClient();

    HttpRequest httpRequest = HttpRequest.newBuilder()
      .uri(URI.create("https://httpstate.com/" + uuid))
      .GET()
      .build();

    HttpResponse<String> httpResponse = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString());

    return httpResponse.statusCode() == 200
      ? httpResponse.body()
      : null;
  }

  public static class Message {
    static class MessageType {
      String uuid;
      long timestamp;
      byte type;
      byte[] value;

      MessageType(String uuid, long timestamp, byte type, byte[] value) {
        this.uuid = uuid;
        this.timestamp = timestamp;
        this.type = type;
        this.value = value;
      }
    }

    public static MessageType unpack(byte[] b) {
      int length = b[0]&0xFF;

      String uuid = new String(b, 1, length, StandardCharsets.UTF_8);
      long timestamp = ByteBuffer.wrap(b, 1+length, 8).getLong();
      byte type = b[1+length+8];
      byte[] value = Arrays.copyOfRange(b, 1+length+9, b.length);

      return new MessageType(uuid, timestamp, type, value);
    }
  }

  static Integer Post(String uuid, String body) throws Exception {
    return Set(uuid, body);
  }

  static Integer Put(String uuid, String body) throws Exception {
    return Set(uuid, body);
  }

  static String Read(String uuid) throws Exception {
    return Get(uuid);
  }

  static Integer Set(String uuid, String body) throws Exception {
    HttpClient httpClient = HttpClient.newHttpClient();

    HttpRequest httpRequest = HttpRequest.newBuilder()
      .uri(URI.create("https://httpstate.com/" + uuid))
      .header("Content-Type", "text/plain;charset=UTF-8")
      .POST(HttpRequest.BodyPublishers.ofString(body))
      .build();

    HttpResponse<String> httpResponse = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString());

    return httpResponse.statusCode();
  }

  static Integer Write(String uuid, String body) throws Exception {
    return Set(uuid, body);
  }

  public String data;
  private final Map<String, List<Consumer<String>>> et;
  private String uuid;
  private WebSocket ws;

  public void Emit(String type) {
    this.Emit(type, null);
  }

  public void Emit(String type, String data) {
    if(this.et.get(type) != null)
      for(Consumer<String> callback : this.et.get(type))
        callback.accept(data);
  }

  public String Get() throws Exception {
    if(this.uuid != null) {
      String data = HttpState.Get(this.uuid);

      if(!data.equals(this.data))
        CompletableFuture.runAsync(() -> this.Emit("change", this.data));

      this.data = data;

      return this.data;
    } else return null;
  }

  public void Off(String type, Consumer<String> callback) {
    if(this.et.get(type) != null) {
      if(callback != null)
        this.et.get(type).remove(callback);

      if(callback == null || this.et.get(type).isEmpty())
        this.et.remove(type);
    }
  }

  public void On(String type, Consumer<String> callback) {
    if(this.et.get(type) == null)
      this.et.put(type, new ArrayList<>());

    this.et.get(type).add(callback);
  }

  public Integer Post(String body) throws Exception {
    return this.Set(body);
  }

  public Integer Put(String body) throws Exception {
    return this.Set(body);
  }

  public String Read() throws Exception {
    return this.Get();
  }

  public Integer Set(String body) throws Exception {
    return this.uuid != null
      ? HttpState.Set(this.uuid, body)
      : null;
  }

  public Integer Write(String body) throws Exception {
    return this.Set(body);
  }

  public HttpState(String uuid) throws Exception {
    this.data = null;
    this.et = new HashMap<String, List<Consumer<String>>>();
    this.uuid = uuid;
    this.ws = null;

    Listener listener = new Listener() {
      @Override
      public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
        System.out.println("onClose");

        return null;
      }

      @Override
      public void onError(WebSocket webSocket, Throwable error) {
        error.printStackTrace();
      }

      @Override
      public void onOpen(WebSocket webSocket) {
        HttpState.this.ws = webSocket;

        HttpState.this.ws.sendText("{\"open\":\"" + HttpState.this.uuid + "\"}", true);

        HttpState.this.Emit("open");

        Executors.newSingleThreadScheduledExecutor()
          .scheduleAtFixedRate(() -> HttpState.this.ws.sendPing(ByteBuffer.allocate(0)), 0, 30, TimeUnit.SECONDS);

        HttpState.this.ws.request(1);
      }

      @Override
      public CompletionStage<?> onBinary(WebSocket webSocket, ByteBuffer _data, boolean last) {
        byte[] bytes = new byte[_data.remaining()];
        _data.get(bytes);

        Message.MessageType data = Message.unpack(bytes);

        if(
             data != null
          && data.uuid.equals(HttpState.this.uuid)
          && data.type == 1
        ) {
          HttpState.this.data = new String(data.value, StandardCharsets.UTF_8);

          HttpState.this.Emit("change", HttpState.this.data);
        }

        HttpState.this.ws.request(1);

        return null;
      }
    };

    HttpClient.newHttpClient()
      .newWebSocketBuilder()
      .buildAsync(URI.create("wss://httpstate.com/" + this.uuid), listener)
      .join();

    this.Get();
  }
}
